### PR TITLE
refactor: benchmarks

### PR DESCRIPTION
**Pull request by McPatate**
Based on a discussion with @LysandreJik & @ArthurZucker, the goal of this PR is to improve transformers' benchmark system.

This is a WIP, for the moment the infrastructure required to make things work is not ready. Will update the PR description when it is the case.

### CODE REVIEW

File: benchmark/llama.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        model.generate(**inputs, do_sample=False)
+        end = perf_counter()
+        first_eager_fwd_pass_time = end - start
+        # TODO:
+        # response = tokenizer.batch_decode(outputs)[0]
+        start = perf_counter()
+        model.generate(**inputs, do_sample=False)
+        end = perf_counter()
+        second_eager_fwd_pass_time = end - start
+
```
Comment (ArthurZucker) [inline benchmark/llama.py:115]:
```suggestion
        torch.compiler.reset() # reset compiler
```

File: benchmark/llama.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        tokenizer = AutoTokenizer.from_pretrained(ckpt)
+
+        prompt = "Why dogs are so cute?"
+        inputs = tokenizer(prompt, return_tensors="pt").to(device)
+
+        # Specify the max length (including both the prompt and the response)
+        # When calling `generate` with `cache_implementation="static" later, this is also used to create a `StaticCache` object
+        # with sequence length = `max_length`. The longer the more you will re-use it
+        model.generation_config.max_length = 128
+
```
Comment (ArthurZucker) [inline benchmark/llama.py:103]: We should also use  (to make sure we are not sampling):
```python
from transformers import GenerationConfig
gen_config = GenerationConfig(do_sample=False, top_p =1, temperature=1)
model = AutoModelForCausalLM.from_pretrained(ckpt, torch_dtype=torch.float16, generation_config = gen_config)
```

File: benchmark/llama.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        # `torch.compile(model, ...)` is not recommended as you compile callbacks
+        # and full generate. We recommend compiling only the forward for now.
+        # "reduce-overhead" will use cudagraphs.
+        model.forward = torch.compile(model.forward, mode="reduce-overhead", fullgraph=True)
+        model.generation_config.cache_implementation = "static"
+
+        # with `torch.compile` (on A100 80G + torch 2.3)
+        # 1st call: ~ 90 seconds
+        start = perf_counter()
+        model.generate(**inputs, do_sample=False)
```
Comment (ArthurZucker) [inline benchmark/llama.py:125]: okay, we are more computing the full generation here ! let's keep this . You can copy past the entire `run_benchmark` and use this: https://gist.github.com/ArthurZucker/5dc54a3fb443e979fac437e5df7c800b 
which is the proper forward calling!

**Review by ArthurZucker — COMMENTED** [review]
🚀

**Review by LysandreJik — APPROVED** [review]
Awesome, let's ship a first version

File: benchmark/llama.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+        # with sequence length = `max_length`. The longer the more you will re-use it
+        seq_length = inputs["input_ids"].shape[1]
+        model.generation_config.max_length = seq_length + num_tokens_to_generate
+        batch_size = inputs["input_ids"].shape[0]
+
+        #########
+        # Eager #
+        #########
+
+        past_key_values = StaticCache(model.config, batch_size=batch_size)
```
Comment (ArthurZucker) [inline benchmark/llama.py:127]:
```suggestion
        past_key_values = StaticCache(model.config, batch_size=batch_size, device=device)
```

File: benchmark/llama.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+        )
+        end = perf_counter()
+        second_eager_fwd_pass_time = end - start
+        logger.info(f"completed second eager fwd pass in: {second_eager_fwd_pass_time}s")
+        start = perf_counter()
+        model.generate(**inputs, do_sample=False)
+        end = perf_counter()
+        second_eager_generate_time = end - start
+        logger.info(f"completed second eager generation in: {second_eager_generate_time}s")
+        logger.info(f"generated: {tokenizer.batch_decode(output)}")
```
Comment (ArthurZucker) [inline benchmark/llama.py:165]:
```suggestion
        logger.info(f"generated: {tokenizer.batch_decode(output.cpu().tolist())}")
```

File: benchmark/llama.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        )
+        end = perf_counter()
+        first_eager_fwd_pass_time = end - start
+        logger.info(f"completed first eager fwd pass in: {first_eager_fwd_pass_time}s")
+        start = perf_counter()
+        output = model.generate(**inputs, do_sample=False)
+        end = perf_counter()
+        first_eager_generate_time = end - start
+        logger.info(f"completed first eager generation in: {first_eager_generate_time}s")
+        logger.info(f"generated: {tokenizer.batch_decode(output)}")
```
Comment (ArthurZucker) [inline benchmark/llama.py:145]:
```suggestion
        logger.info(f"generated: {tokenizer.batch_decode(output.cpu().tolist())}")
```

**Comment by ArthurZucker** [comment]
Compile complains IDK why, will handle, rest mr @McPatate

**Review by ArthurZucker — APPROVED** [review]

**Pull request merged**